### PR TITLE
[guilib] Use binary search for letter navigation in GUIBaseContainer

### PR DESCRIPTION
**Pull request by pannal**
## Description
Replace linear O(n) searches in OnNextLetter(), OnPrevLetter(), and
OnJumpSMS() with binary search O(log n) using std::upper_bound and
std::lower_bound.

The m_letterOffsets vector is already sorted by offset, making it
ideal for binary search. This improves letter jump performance on
large lists (500+ items) from ~250 comparisons average to ~9.

## Motivation and context
Performance

## How has this been tested?
Kodi 22 master on Windows, CoreELEC 21.3 p3i on Ugoos AM6B+

## What is the effect on users?
Better performance

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **Student submission** (PR was done for educational purposes and will be treated as such)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by pannal** [comment]
1. upper_bound returns begin() edge case: When offset is before all letter positions, it stays at begin() (the decrement is skipped). Dereferencing begin() gives the first letter - which is the same behavior as the original code.  The original's while loop would stop at currentLetter = 0, also giving the first element. Behavior is identical. 
2. letters.find(current) returning npos: This is a pre-existing behavior, identical in the original code at the same line (size_t startPos = (letters.find(current) + 1) % letters.size()). The commit didn't change this logic at all. And in practice, npos + 1 wraps to 0, so startPos becomes 0 % letters.size() = 0, which cycles through all letters from the start - a reasonable fallback when the current letter isn't in this SMS key's group.

The bot is confused by its own reasoning (it even contradicts itself mid-comment: "Actually, the critical bug is... the current logic leaves it pointing to begin() which is correct").

File: xbmc/guilib/GUIBaseContainer.cpp [hunk 1]
```diff
@@ -630,30 +631,26 @@ void CGUIBaseContainer::OnRight()
...
-  {
-    if (m_letterOffsets[i].first > offset)
-    {
-      SelectItem(m_letterOffsets[i].first);
-      return;
-    }
-  }
+  const int offset = CorrectOffset(GetOffset(), GetCursor());
+  // Binary search for first letter offset greater than current position
+  auto it = std::upper_bound(m_letterOffsets.begin(), m_letterOffsets.end(), offset,
```
Comment (sundermann) [inline xbmc/guilib/GUIBaseContainer.cpp]: Please use ranges (and projection would also be possible here):
```suggestion
  auto it = std::ranges::upper_bound(m_letterOffsets, offset, {}, &std::pair<int, std::string>::first);
```

Same for the other calls below. You don't need `<ranges>`, `<algorithm>` is enough

Comment (pannal) [reply xbmc/guilib/GUIBaseContainer.cpp]: Thanks, I'll take a look!

**Review by neo1973 — APPROVED** [review]

**Comment by garbear** [comment]
@claude please review

**Review by garbear — APPROVED** [review]
